tsinfer/inference.py

```diff
@@ -890,23 +980,77 @@ def __init__(
         )
         self.position_map[0] = 0
+        self.recombination = np.zeros(self.num_sites)  # TODO: reduce len by 1
+        self.mismatch = np.zeros(self.num_sites)
+
+        # Set the canonical 'recombination' and 'mismatch' arrays to appropriate values
+        # NB 'recombination' must be of length num_sites-1, mismatch of length num_sites
+
+        # First set defaults if None provided
+        if recombination_rate is None and recombination is None:
+            if mismatch_ratio is None and mismatch is None:
+                # Special case: revert to tsinfer 0.1 behaviour with no mismatch allowed
+                mismatch = np.full(self.num_sites, 1e-20)  # A miniscule value
+                # arbitrary, large relative to mismatch, but < 1
+                if self.num_sites == 0:
+                    recombination = []
+                else:
+                    recombination = np.full(self.num_sites - 1, 1e-2)
+                # Shouldn't need insane precision, although 3 is not enough in trials
+                if precision is None:
+                    precision = 5
+            else:
+                raise ValueError("Cannot use mismatch without setting recombination")
+        if mismatch_ratio is None and mismatch is None:
+            # Default if recombination but no mismatch params have been specified
+            mismatch_ratio = 1.0
+
+        assert not (recombination_rate is None and recombination is None)
+        assert not (mismatch_ratio is None and mismatch is None)
+
+        # Set recombination from recombination_rate
+        if recombination is None:
+            try:
+                recombination = np.diff(
+                    recombination_rate.get_cumulative_mass(inference_site_position)
+                )
+            except AttributeError:
+                recombination = np.diff(inference_site_position * recombination_rate)
+        elif recombination_rate is not None:
+            raise ValueError(
+                "Cannot specify both recombination rate and recombination probabilities"
+            )
+        if len(recombination) != max(0, self.num_sites - 1):
+            raise ValueError("Bad length for recombination array")
+
+        # Set mismatch from mismatch_ratio
+        if mismatch is None:
+            if len(recombination) > 0:
+                mismatch = np.full_like(
+                    self.mismatch, mismatch_ratio * np.median(recombination)
+                )
+            else:
+                # Hack if < 2 sites: https://github.com/tskit-dev/tsinfer/issues/260
+                mismatch = np.full_like(self.mismatch, 0.1)
+        elif mismatch_ratio is not None:
+            raise ValueError(
+                "Cannot specify both a mismatch ratio and mismatch probabilities"
+            )
+        if len(mismatch) != self.num_sites:
+            raise ValueError("Bad length for mismatch array")
 
         if precision is None:
+            nonzero_probs = np.concatenate(
```

Comment:
Again, let's separate out precision from this update. It's not necessary and we want to change the minimum number of things we can when making fundamental changes like this.